esp/esdllib/esdl_script.cpp

```diff
@@ -0,0 +1,914 @@
+    {
+        if (isEmptyString(tree->queryProp("@name")))
+            esdlOperationError("Custom transform: ", "Variable without name", "", !ignoreCodingErrors);
+        m_name.set(tree->queryProp("@name"));
+        if (isEmptyString(tree->queryProp("@select")))
+            esdlOperationError("Custom transform: ", "Variable without select", m_name.str(), !ignoreCodingErrors);
+        if (tree->hasProp("@select"))
+            m_select.setown(compileXpath(tree->queryProp("@select")));
+    }
+
+    virtual ~CEsdlTransformOperationVariable()
+    {
+    }
+
+    virtual bool process(IEspContext * context, IPropertyTree *request, IXpathContext * xpathContext) override
+    {
+        return xpathContext->addCompiledVariable(m_name, m_select);
+    }
+
+    virtual void toDBGLog() override
+    {
+#if defined(_DEBUG)
+        DBGLOG(">%s> %s with select(%s)", m_name.str(), tagname.str(), m_select.get() ? m_select->getXpath() : "");
+#endif
+    }
+};
+
+class CEsdlTransformOperationParameter : public CEsdlTransformOperationBase
+{
+protected:
+    StringAttr m_name;
+    Owned<ICompiledXpath> m_select;
+
+public:
+    CEsdlTransformOperationParameter(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationBase(tree, prefix)
+    {
+        m_name.set(tree->queryProp("@name"));
+        if (tree->hasProp("@select"))
+            m_select.setown(compileXpath(tree->queryProp("@select")));
+    }
+
+    virtual ~CEsdlTransformOperationParameter()
+    {
+    }
+
+    virtual bool process(IEspContext * context, IPropertyTree *request, IXpathContext * xpathContext) override
+    {
+        return xpathContext->declareCompiledParameter(m_name, m_select);
+    }
+
+    virtual void toDBGLog() override
+    {
+#if defined(_DEBUG)
+        DBGLOG(">%s> %s with select(%s)", m_name.str(), tagname.str(), m_select.get() ? m_select->getXpath() : "");
+#endif
+    }
+};
+
+class CEsdlTransformOperationForEach : public CEsdlTransformOperationBase
+{
+protected:
+    Owned<ICompiledXpath> m_select;
+    IArrayOf<IEsdlTransformOperation> m_variables;
+    IArrayOf<IEsdlTransformOperation> m_children;
+
+public:
+    CEsdlTransformOperationForEach(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationBase(tree, prefix)
+    {
+        if (tree)
+        {
+            if (isEmptyString(tree->queryProp("@select")))
+                esdlOperationError("Custom transform: ", "for-each without select", "", !ignoreCodingErrors);
+            m_select.setown(compileXpath(tree->queryProp("@select")));
+
+            loadChildren(tree, prefix);
+        }
+    }
+
+    virtual ~CEsdlTransformOperationForEach(){}
+
+    bool process(IEspContext * context, IPropertyTree *request, IXpathContext * xpathContext) override
+    {
+        Owned<IXpathContextIterator> contexts = evaluate(xpathContext);
+        if (!contexts)
+            return false;
```

Comment:
Could do with more thought.  False doesn't mean failure... it means more like didn't match or didn't do anything.  It's most useful in conditionals and especially inside a choose.  But I'll review the consistency of the returns.

Reply:
I suppose I'm influenced by my refactoring. When I process statements/operations, I'm more concerned with whether something prevented a statement from completing its requested task. Whether a choose finds a matching child doesn't matter to my engine, but the engine absolutely wants to know if an evaluation failed to produce a usable result. So when I saw bool returned without seeing how it might be used, I assumed it meant the processing did or did not complete without error.